File: python_test_crate/tests/15_system_return_chain.py
from typing import Any, Optional, List, Dict, Callable

class SystemReturnChainTestFrameEvent:
    def __init__(self, message: str, parameters):
        self._message = message
        self._parameters = parameters
        self._return = None


class SystemReturnChainTestCompartment:
    def __init__(self, state: str, parent_compartment = None):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment

    def copy(self) -> 'SystemReturnChainTestCompartment':
        c = SystemReturnChainTestCompartment(self.state, self.parent_compartment)
        c.state_args = self.state_args.copy()
        c.state_vars = self.state_vars.copy()
        c.enter_args = self.enter_args.copy()
        c.exit_args = self.exit_args.copy()
        c.forward_event = self.forward_event
        return c


class SystemReturnChainTest:
    def __init__(self):
        self._state_stack = []
        self._return_value = None
        self.__compartment = SystemReturnChainTestCompartment("Start")
        self.__next_compartment = None
        __frame_event = SystemReturnChainTestFrameEvent("$>", None)
        self.__kernel(__frame_event)

    def __kernel(self, __e):
        # Route event to current state
        self.__router(__e)
        # Process any pending transition
        while self.__next_compartment is not None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            # Exit current state
            exit_event = SystemReturnChainTestFrameEvent("<$", self.__compartment.exit_args)
            self.__router(exit_event)
            # Switch to new compartment
            self.__compartment = next_compartment
            # Enter new state (or forward event)
            if next_compartment.forward_event is None:
                enter_event = SystemReturnChainTestFrameEvent("$>", self.__compartment.enter_args)
                self.__router(enter_event)
            else:
                # Forward event to new state
                forward_event = next_compartment.forward_event
                next_compartment.forward_event = None
                if forward_event._message == "$>":
                    # Forwarding enter event - just send it
                    self.__router(forward_event)
                else:
                    # Forwarding other event - send $> first, then forward
                    enter_event = SystemReturnChainTestFrameEvent("$>", self.__compartment.enter_args)
                    self.__router(enter_event)
                    self.__router(forward_event)

    def __router(self, __e):
        state_name = self.__compartment.state
        handler_name = f"_state_{state_name}"
        handler = getattr(self, handler_name, None)
        if handler:
            handler(__e)

    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment

    def test_enter_sets(self) -> str:
        self._return_value = None
        __e = SystemReturnChainTestFrameEvent("test_enter_sets", None)
        self.__kernel(__e)
        return self._return_value

    def test_exit_then_enter(self) -> str:
        self._return_value = None
        __e = SystemReturnChainTestFrameEvent("test_exit_then_enter", None)
        self.__kernel(__e)
        return self._return_value

    def get_state(self) -> str:
        self._return_value = None
        __e = SystemReturnChainTestFrameEvent("get_state", None)
        self.__kernel(__e)
        return self._return_value

    def _state_EnterSetter(self, __e):
        if __e._message == "$>":
            # Enter handler sets return value
            self._return_value = "from_enter"
        elif __e._message == "get_state":
            self._return_value = "EnterSetter"
            __e._return = self._return_value
            return

    def _state_BothSet(self, __e):
        if __e._message == "$>":
            # Enter handler sets return - should overwrite exit's value
            self._return_value = "enter_wins"
        elif __e._message == "get_state":
            self._return_value = "BothSet"
            __e._return = self._return_value
            return

    def _state_Start(self, __e):
        if __e._message == "<$":
            # Exit handler sets initial value
            self._return_value = "from_exit"
        elif __e._message == "get_state":
            self._return_value = "Start"
            __e._return = self._return_value
            return
        elif __e._message == "test_enter_sets":
            __compartment = SystemReturnChainTestCompartment("EnterSetter")
            self.__transition(__compartment)
        elif __e._message == "test_exit_then_enter":
            __compartment = SystemReturnChainTestCompartment("BothSet")
            self.__transition(__compartment)


def main():
    print("=== Test 15: System Return Chain (Last Writer Wins) ===")

    # Test 1: Start exit + EnterSetter enter
    # Start's exit sets "from_exit", EnterSetter's enter sets "from_enter"
    # Enter should win (last writer)
    s1 = SystemReturnChainTest()
    result = s1.test_enter_sets()
    assert result == "from_enter", f"Expected 'from_enter', got '{result}'"
    assert s1.get_state() == "EnterSetter", f"Expected state 'EnterSetter'"
    print(f"1. Exit set then enter set - enter wins: '{result}'")

    # Test 2: Both handlers set, enter wins
    s2 = SystemReturnChainTest()
    result = s2.test_exit_then_enter()
    assert result == "enter_wins", f"Expected 'enter_wins', got '{result}'"
    assert s2.get_state() == "BothSet", f"Expected state 'BothSet'"
    print(f"2. Both set - enter wins: '{result}'")

    print("PASS: System return chain (last writer wins) works correctly")

if __name__ == '__main__':
    main()

File: typescript_test_crate/tests/15_system_return_chain.ts

// Tests that system.return follows "last writer wins" across transition lifecycle


class SystemReturnChainTestFrameEvent {
    public _message: string;
    public _parameters: Record<string, any> | null;
    public _return: any;

    constructor(message: string, parameters: Record<string, any> | null) {
        this._message = message;
        this._parameters = parameters;
        this._return = null;
    }
}


class SystemReturnChainTestCompartment {
    public state: string;
    public state_args: Record<string, any>;
    public state_vars: Record<string, any>;
    public enter_args: Record<string, any>;
    public exit_args: Record<string, any>;
    public forward_event: any;
    public parent_compartment: SystemReturnChainTestCompartment | null;

    constructor(state: string, parent_compartment: SystemReturnChainTestCompartment | null = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    public copy(): SystemReturnChainTestCompartment {
        const c = new SystemReturnChainTestCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


class SystemReturnChainTest {
    private _state_stack: Array<any>;
    private __compartment: SystemReturnChainTestCompartment;
    private __next_compartment: SystemReturnChainTestCompartment | null;
    private _return_value: any;

    constructor() {
        this._state_stack = [];
        this._return_value = null;
        this.__compartment = new SystemReturnChainTestCompartment("Start");
        this.__next_compartment = null;
        const __frame_event = new SystemReturnChainTestFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    private __kernel(__e: SystemReturnChainTestFrameEvent) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new SystemReturnChainTestFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new SystemReturnChainTestFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new SystemReturnChainTestFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    private __router(__e: SystemReturnChainTestFrameEvent) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = (this as any)[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    private __transition(next_compartment: SystemReturnChainTestCompartment) {
        this.__next_compartment = next_compartment;
    }

    public test_enter_sets(): string {
        this._return_value = null;
        const __e = new SystemReturnChainTestFrameEvent("test_enter_sets", null);
        this.__kernel(__e);
        return this._return_value;
    }

    public test_exit_then_enter(): string {
        this._return_value = null;
        const __e = new SystemReturnChainTestFrameEvent("test_exit_then_enter", null);
        this.__kernel(__e);
        return this._return_value;
    }

    public get_state(): string {
        this._return_value = null;
        const __e = new SystemReturnChainTestFrameEvent("get_state", null);
        this.__kernel(__e);
        return this._return_value;
    }

    private _state_BothSet(__e: SystemReturnChainTestFrameEvent) {
        if (__e._message === "$>") {
            // Enter handler sets return - should overwrite exit's value
            this._return_value = "enter_wins";;
        } else if (__e._message === "get_state") {
            this._return_value = "BothSet";
            __e._return = this._return_value;
            return;;
        }
    }

    private _state_Start(__e: SystemReturnChainTestFrameEvent) {
        if (__e._message === "<$") {
            // Exit handler sets initial value
            this._return_value = "from_exit";;
        } else if (__e._message === "get_state") {
            this._return_value = "Start";
            __e._return = this._return_value;
            return;;
        } else if (__e._message === "test_enter_sets") {
            const __compartment = new SystemReturnChainTestCompartment("EnterSetter", this.__compartment.copy());
            this.__transition(__compartment);
        } else if (__e._message === "test_exit_then_enter") {
            const __compartment = new SystemReturnChainTestCompartment("BothSet", this.__compartment.copy());
            this.__transition(__compartment);
        }
    }

    private _state_EnterSetter(__e: SystemReturnChainTestFrameEvent) {
        if (__e._message === "$>") {
            // Enter handler sets return value
            this._return_value = "from_enter";;
        } else if (__e._message === "get_state") {
            this._return_value = "EnterSetter";
            __e._return = this._return_value;
            return;;
        }
    }
}


function main() {
    console.log("=== Test 15: System Return Chain (Last Writer Wins) ===");

    // Test 1: Start exit + EnterSetter enter
    // Start's exit sets "from_exit", EnterSetter's enter sets "from_enter"
    // Enter should win (last writer)
    const s1 = new SystemReturnChainTest();
    const result1 = s1.test_enter_sets();
    if (result1 !== "from_enter") {
        throw new Error(`Expected 'from_enter', got '${result1}'`);
    }
    if (s1.get_state() !== "EnterSetter") {
        throw new Error(`Expected state 'EnterSetter'`);
    }
    console.log(`1. Exit set then enter set - enter wins: '${result1}'`);

    // Test 2: Both handlers set, enter wins
    const s2 = new SystemReturnChainTest();
    const result2 = s2.test_exit_then_enter();
    if (result2 !== "enter_wins") {
        throw new Error(`Expected 'enter_wins', got '${result2}'`);
    }
    if (s2.get_state() !== "BothSet") {
        throw new Error(`Expected state 'BothSet'`);
    }
    console.log(`2. Both set - enter wins: '${result2}'`);

    console.log("PASS: System return chain (last writer wins) works correctly");
}

main();
